proposals/xxxx-initializer-lists.md

```diff
@@ -0,0 +1,169 @@
+<!-- {% raw %} -->
+
+# Initializer Lists
+
+* Proposal: [NNNN](NNNN-initializer-lists.md)
+* Author(s): [Chris Bieneman](https://github.com/llvm-beanz)
+* Status: **Design In Progress**
+
+
+## Introduction
+
+HLSL's initializer lists assume a member-by-member flattened initialization
+sequence. It is significantly different in behavior from C/C++ initializer
+lists, but in a way that may not be apparent to users that they are depending on
+the divergent behavior. This proposal suggests an approach to implementing
+HLSL's initializer lists with a forward-looking perspective.
+
+## Motivation
+
+As an example, the initializer lists in the code below are all valid in HLSL today:
+
+```hlsl
+struct A {
+  int a;
+  double b;
+};
+struct B {
+  A a[2];
+  int c;
+};
+B b = {{1, 1.2}, {2, 2.2}, 3};   // Array elements specified as members
+B b2 = {1, 2, 3, 4, 5};          // each field initialized separately
+B b3 = {{1, {2, 3}}, {4, 5}};    // Completely random grouping of arguments
+int4 i4 = {1,2,3,4};             // valid int4 in C-syntax
+B b4 = {i4, 5};                  // int4 implicitly expanded to 4 arguments
+```
+
+In HLSL, initializer arguments are scalarized, and a member-by-member
+initialization is performed in depth-first order from the initializer list to
+the destination type. All scalarized fields in a destination type must map to a
+scalarized field from an initializer argument, and the argument must be
+convertible to the field type. It is an error if the source or destination type
+have a different number of scalarized components when fully flattened.
+
+
+In C++, initializer lists must represent the structure of the object they are
+initializing, and each initializer in the list must be convertible to the
+corresponding argument, however not all fields must be initialized. If the
+initializer has less arguments than the target has fields, the remaining fields
+are zero initialized. It is an error if an initializer list has too many
+arguments, if an initializer has arguments that cannot be converted to the
+destination type, or if an initializer list has incorrect nested braces.
+
+
+The following examples which are valid in C++ are not valid in HLSL:
+
+```c++
+struct A {
+  int a;
+  double b;
+};
+struct B {
+  A a[2];
+  int c;
+};
+B b = {};
+A a = {0};
+B b2 = {{{0}}, 0};
+B b3 = {{0}, 0};
+```
+
+Given the differences in the language described above it is likely too
+significant of a change to make Clang follow C/C++ behavior without a transition
+period. Since Clang and DXC will have different underlying representations for
```

Comment:
```suggestion
Given the differences in the languages as described above it is likely changes
needed to make Clang follow C/C++ behavior would be significant thereby requiring
a transition period. Since Clang and DXC will have different underlying representations for
```